### PR TITLE
Android compatibility

### PR DESCRIPTION
**Pull request by Sisah2**
Will work out of the box on android, but need latest gl4es version.

### CODE REVIEW

**Comment by Sisah2** [comment]
Can put cloud_transition and mist_transition to p4evious place, but it will need change it to a function, same as weather_tone.

**Comment by zesterer** [comment]
Was this closed on purpose?

**Comment by Sisah2** [comment]
Yes, there is something wrong with mist brightness it seems, i reopen this when i figure it out.

**Comment by zesterer** [comment]
FWIW you can just push new commits while you're figuring it out, no need to close it. If it's still in development you should be able to mark it as a draft PR.

**Comment by Sisah2** [comment]
So found the reason for brighter/different mist brightness, it started with apply_fog changes in this commit
https://github.com/zesterer/openmw-volumetric-clouds/commit/a8627a82c46b2fea3008f278b18860cfa5e6c0f3
So can reopen.

File: shaders/clouds.omwfx [hunk 1]
```diff
@@ -472,33 +449,33 @@ fragment main {
...
+        if( i == 1) return vec3(0.7); //Cloudy
+        if( i == 2) return vec3(0.65);  //Foggy
+        if( i == 3) return vec3(0.6);  //Overcast
+        if( i == 4) return vec3(0.55);  //Rain
+        if( i == 5) return vec3(0.5, 0.5, 0.6);  //Thunder
+        if( i == 6) return vec3(1.0, 0.5, 0.25);  //Ash
+        if( i == 7) return vec3(1.0, 0.5, 0.5);  //Blight
+        if( i == 8) return vec3(1.0);   //Snow
+        return vec3(1.0);   //Blizzard
+    }
```
Comment (zesterer) [inline shaders/clouds.omwfx:468]: I am slightly nervous about how well this optimises in practice. That said, the rest of the shader is substantially more expensive, so I guess it's fine.

File: shaders/clouds.omwfx [hunk 2]
```diff
@@ -324,10 +295,16 @@ fragment main {
...
     #define M3 396128171
 
-    // Replace this function with the one below it if your GPU can't handle bitwise ops
+    // Set to 1 if your GPU can't handle bitwise ops
+    #define NO_BITWISE 0
+
     float hash_one(int q) {
+#if defined(GL4ES) || NO_BITWISE
+        return fract(sin(float(q)) * 43758.5453123);
+#else
```
Comment (zesterer) [inline shaders/clouds.omwfx:304]: Nice :)

File: shaders/clouds.omwfx [hunk 3]
```diff
@@ -637,6 +614,33 @@ fragment main {
...
+        mist_transition[1] = float(0.25);  //Cloudy
+        mist_transition[2] = float(0.75);  //Foggy
+        mist_transition[3] = float(0.35);  //Overcast
+        mist_transition[4] = float(0.4);  //Rain
+        mist_transition[5] = float(0.4);  //Thunder
+        mist_transition[6] = float(0.3);  //Ash
+        mist_transition[7] = float(0.2);  //Blight
+        mist_transition[8] = float(0.3);  //Snow
+        mist_transition[9] = float(0.4);   //Blizzard
+
```
Comment (zesterer) [inline shaders/clouds.omwfx:643]: Why not put these into dedicated functions too like `weather_tone`? At least then the degree to which the compiler can optimise things is consistent for each GL impl.

**Review by zesterer — CHANGES_REQUESTED** [review]
Broadly very happy with this PR, just one change. Thanks for opening!

File: shaders/clouds.omwfx (hunk 3 above)
Comment (Sisah2) [reply shaders/clouds.omwfx]: Ok, changed them to a function.

**Comment by zesterer** [comment]
Thanks for the PR, much appreciated!

**Pull request merged**